// eureka-client/lib/Logger.js
'use strict';

exports.__esModule = true;

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var LEVELS = {
  error: 50,
  warn: 40,
  info: 30,
  debug: 20
};
var DEFAULT_LEVEL = LEVELS.info;

var Logger = function () {
  function Logger() {
    _classCallCheck(this, Logger);

    this._level = DEFAULT_LEVEL;
  }

  Logger.prototype.level = function level(inVal) {
    var val = inVal;
    if (val) {
      if (typeof val === 'string') {
        val = LEVELS[val];
      }
      this._level = val || DEFAULT_LEVEL;
    }
    return this._level;
  };

  // Abstract the console call:

  Logger.prototype._log = function _log(method, args) {
    if (this._level <= LEVELS[method === 'log' ? 'debug' : method]) {
      var _console;

      /* eslint-disable no-console */
      (_console = console)[method].apply(_console, args);
      /* eslint-enable no-console */
    }
  };

  Logger.prototype.error = function error() {
    return this._log('error', arguments);
  };

  Logger.prototype.warn = function warn() {
    return this._log('warn', arguments);
  };

  Logger.prototype.info = function info() {
    return this._log('info', arguments);
  };

  Logger.prototype.debug = function debug() {
    return this._log('log', arguments);
  };

  return Logger;
}();

exports.default = Logger;
